rules/detect-non-literal-fs-filename.js

```diff
@@ -5,12 +5,212 @@
 
 'use strict';
 
+const fsMetaData = require('./data/fsFunctionData.json');
+const funcNames = Object.keys(fsMetaData);
+const fsPackageNames = require('./data/fsPackagesNames.json');
+const { getImportDeclaration, getVariableDeclaration } = require('./utils/import-utils');
+
 //------------------------------------------------------------------------------
-// Rule Definition
+// Utils
 //------------------------------------------------------------------------------
 
-const fsMetaData = require('./data/fsFunctionData.json');
-const funcNames = Object.keys(fsMetaData);
+function sinkPositions(node, argMeta) {
+  return (argMeta || []).filter((argIndex) => node.arguments[argIndex].type !== 'Literal');
+}
+
+function generateReport({ context, node, packageName, methodName, indeces }) {
+  if (!indeces || indeces.length === 0) {
+    return null;
+  }
+  return context.report(node, `Found ${methodName} from package "${packageName}" with non literal argument at index ${indeces.join(',')}`);
+}
+
+/**
+ * Detects:
+ * | var something = require('fs').readFile;
+ * | something(a);
+ */
+function sinksForRequireWithProperty({ context, methodName, node, program }) {
+  const declaration = getVariableDeclaration({
+    condition: (declaration) => declaration.init.parent.id.name === methodName,
+    hasObject: true,
+    methodName,
+    packageNames: fsPackageNames,
+    program,
+  });
+
+  if (!declaration) {
+    return null;
+  }
+
+  // we found the require for our method!
+  const fsFunction = declaration.init.property.name;
+  const packageName = declaration.init.object.arguments[0].value;
+  const fnName = declaration.init.property.name;
+
+  const sinks = sinkPositions(node, fsMetaData[fsFunction]);
+
+  return generateReport({
+    context,
+    node,
+    packageName,
+    methodName: fnName,
+    indeces: sinks,
+  });
+}
+
+/**
+ * Detects:
+ * | var something = require('fs');
+ * | something.readFile(c);
+ */
+function sinkForMethodCall({ context, node, program, methodName }) {
+  const declaration = getVariableDeclaration({
+    packageNames: fsPackageNames,
+    hasObject: false,
+    program,
+  });
+
+  if (!declaration) {
+    return null;
+  }
+
+  const sinks = sinkPositions(node.parent, fsMetaData[methodName]);
```

Comment:
You use the word “sink” throughout, but it’s not clear to me what this means. Can you rename so that it’s clearer?

Reply:
That's a taint analysis terminology (https://owasp.org/www-community/controls/Static_Code_Analysis). But i agree that's not a commonly known term. I tried to rename it.